pkg/skaffold/build/custom/custom.go

```diff
@@ -52,19 +56,57 @@ func NewArtifactBuilder(pushImages bool, additionalEnv []string) *ArtifactBuilde
 // Build builds a custom artifact
 // It returns true if the image is expected to exist remotely, or false if it is expected to exist locally
 func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Artifact, tag string) error {
-	cmd, err := b.retrieveCmd(ctx, out, a, tag)
+	cmd, err := b.retrieveCmd(out, a, tag)
 	if err != nil {
 		return errors.Wrap(err, "retrieving cmd")
 	}
 
-	return cmd.Run()
+	if err := cmd.Start(); err != nil {
+		return errors.Wrap(err, "starting cmd")
+	}
+
+	return b.handleGracefulTermination(ctx, cmd)
+}
+
+func (b *ArtifactBuilder) handleGracefulTermination(ctx context.Context, cmd *exec.Cmd) error {
+	done := make(chan struct{})
```

Comment:
This looks really great. Just one question, i am assuming you have the done channel to not send kill signal if the process is already done? Lets put a defer close as soon as you create the channel

We can also shorten this code block further

```
func (b *ArtifactBuilder) handleGracefulTermination(ctx context.Context, cmd *exec.Cmd) error {
  done := make(chan struct{})
 defer(close(done))
  select {
  case <- ctx.Done():
    // Send interrupt signal if not windows.
  case <- time.After(termination*time.Second):
   // Send kill signal
  case <-done
 }
 err := cmd.Wait()
 return err
}


 }
```

I think this is it. And then we can merge this in!

Thank you for your patience.

Reply:
According to your suggestion, the process will only run for `termination` seconds and then be killed. i think the nested `select` is still necessary to wait for the context to be cancelled...

Reply:
Sounds good.  Running kokoro tests now